src/amp-url-creator.js

```diff
@@ -35,6 +46,55 @@ const HAS_RTL_CHARS = new RegExp('[' + RTL_CHARS + ']');
 /** @private {number} */
 const MAX_DOMAIN_LABEL_LENGTH_ = 63;
 
+
+/**
+ * Constructs a Viewer proxy url using these rules:
+ * https://developers.google.com/amp/cache/overview
+ * 
+ * Example:
+ * Input domain 'http://ampproject.org' can return 
+ * 'https://www-ampproject-org.cdn.ampproject.org/v/s/www.ampproject.org/?amp_js_v=0.1#origin=http%3A%2F%2Flocalhost%3A8000'
+ * 
+ * @param {string} domain The publisher domain
+ * @param {string} protocol The publisher domain
+ * @param {object} initParams Params containing origin, etc.
+ * @param {string} opt_proxyUrlAuthority
+ * @param {string} opt_viewer_js_version
+ * @private
+ */
+export function constructViewerProxyUrl(domain, protocol, initParams,
+  opt_proxyUrlAuthority, opt_viewer_js_version) {
+  const proxyUrl = constructProxyUrl_(domain, opt_proxyUrlAuthority);
+  const protocolStr = protocol == 'https:' ? 's/' : '';
+  const viewerJsVersion = opt_viewer_js_version ? opt_viewer_js_version :
+    DEFAULT_VIEWER_JS_VERSION_;
+  const parsedDomain = parseUrl(domain);
+
+  return proxyUrl + 
+          '/v/' +
+          protocolStr +
+          parsedDomain.host + 
+          '/?amp_js_v=' + viewerJsVersion +
+          '#' +
+          paramsToString_(initParams);
+}
+
+/**
+ * Constructs a proxy url. For example:
+ * 
+ * Input domain 'http://ampproject.org'
+ * will return  'https://www-ampproject-org.cdn.ampproject.org'
+ * 
+ * @param {string} domain The publisher domain
+ * @param {string} opt_proxyUrlAuthority
+ * @private
+ */
+function constructProxyUrl_(domain, opt_proxyUrlAuthority) {
```

Comment:
I would suggest renaming the function too. It returns the origin not the complete cache url.